snuba/pipeline/composite_storage_processing.py

```diff
@@ -0,0 +1,378 @@
+from __future__ import annotations
+
+from typing import Mapping, NamedTuple, Optional, Sequence, Tuple, Union
+
+import sentry_sdk
+
+from snuba.clickhouse.query import Query as ClickhouseQuery
+from snuba.clusters.storage_sets import StorageSetKey, is_valid_storage_set_combination
+from snuba.datasets.plans.query_plan import SubqueryProcessors
+from snuba.datasets.plans.storage_processing import (
+    ClickhouseQueryPlanNew,
+    build_best_plan,
+)
+from snuba.query import ProcessableQuery
+from snuba.query.composite import CompositeQuery
+from snuba.query.data_source.join import IndividualNode, JoinClause, JoinVisitor
+from snuba.query.data_source.simple import Table
+from snuba.query.data_source.visitor import DataSourceVisitor
+from snuba.query.joins.semi_joins import SemiJoinOptimizer
+from snuba.query.processors.physical import ClickhouseQueryProcessor
+from snuba.query.query_settings import QuerySettings
+from snuba.state import explain_meta
+
+
+def apply_composite_storage_processors(
+    query_plan: CompositeDataSourcePlan, settings: QuerySettings
+) -> CompositeQuery[Table]:
+    """
+    Given a composite query plan, this function applies all the
+    db_processors and plan_processors defined on the plan to the
+    physical query.
+    """
+    assert isinstance(query_plan.translated_source, CompositeQuery)
+    plan_root_processors, plan_aliased_processors = query_plan.get_plan_processors()
+    ProcessorsExecutor(
+        plan_root_processors,
+        plan_aliased_processors,
+        settings,
+    ).visit(query_plan.translated_source)
+    root_db_processors, aliased_db_processors = query_plan.get_db_processors()
+    ProcessorsExecutor(root_db_processors, aliased_db_processors, settings).visit(
+        query_plan.translated_source
+    )
+    composite_processors = [SemiJoinOptimizer()]
+    for p in composite_processors:
+        if settings.get_dry_run():
+            with explain_meta.with_query_differ(
+                "composite_storage_processor",
+                type(p).__name__,
+                query_plan.translated_source,
+            ):
+                p.process_query(query_plan.translated_source, settings)
+        else:
+            p.process_query(query_plan.translated_source, settings)
+
+    return query_plan.translated_source
+
+
+def build_best_plan_for_composite_query(
+    physical_query: CompositeQuery[Table],
+    settings: QuerySettings,
+    post_processors: Sequence[ClickhouseQueryProcessor] = [],
+) -> CompositeDataSourcePlan:
+    """
+    A function that traverses each sub-query node in a physical composite query,
+    builds a plan for each sub-query, combines them into a single composite plan, and
+    returns all the necessary storage processors that need to be applied.
+    """
+    plan = CompositeDataSourcePlanner(settings).visit(physical_query.get_from_clause())
+    physical_query.set_from_clause(plan.translated_source)
```

Comment:
its unclear to me why we modify the original query here, i thought we were just trying to find the processors

Reply:
is the query plan responsible for more than just selecting processors?